docs/settings.md
# Settings

All settings are defined through environment variable.

## Global configuration

| Name                     | Type     | Default                    | Comment                            |
| ------------------------ | -------- | -------------------------- | ---------------------------------- |
| VERSION                  | String   | 0.0.0-development          | Application version                |
| APP_PUBLIC_PATH          | String   | /public/                   | Public path for assets             |
| APP_PORT                 | Number   | 3000                       | HTTP server listening port         |
| APP_GZIP                 | Boolean  | true                       | Enable GZip compression            |
| APP_SECURE_COOKIE        | Boolean  | true                       | Enable secure cookies              |
| APP_COOKIE_POLICY        | String   | strict                     | Cookie sameSite policy             |
| APP_DEBUG                | Boolean  | false                      | Enable debug mode                  |
| APP_ENDPOINT             | String   | http://localhost:$APP_PORT | Application endpoint use for links |
| APP_VERBOSE              | Boolean  | true                       | Enable console logs                |
| APP_ALLOWED_DOMAINS      | String[] | []                         | Allowed Domains                    |
| APP_SOCKET_CLIENT_DOMAIN | String   |                            | Socket client domain               |

It's recommended to disable compression if you may delegate it to a reverse proxy.

## Database configuration

| Name            | Type   | Default                       | Comment             |
| --------------- | ------ | ----------------------------- | ------------------- |
| APP_DB_NAME     | String | vorp                          | Database Name       |
| APP_DB_URI      | String | mongodb+srv://localhost:27017 | Database URI String |
| APP_DB_USERNAME | String | root                          | Database Username   |
| APP_DB_PASSWORD | String | mongodb+srv://localhost:27017 | Database Password   |

## Redis configuration

| Name          | Type   | Default                | Comment                 |
| ------------- | ------ | ---------------------- | ----------------------- |
| APP_REDIS_URI | String | redis://127.0.0.1:6379 | Redis connection string |

## Bull configuration

| Name             | Type    | Default | Comment                                        |
| ---------------- | ------- | ------- | ---------------------------------------------- |
| APP_BULL_PERSIST | Boolean | false   | Set default behavior to not drop redis entries |
| APP_BULL_MONITOR | Boolean | false   | Serve bull monitor endpoint on web server      |

Monitor cannot be turned on when running on production environment, that could lead to a leak on sensitive data.
The previous settings are only meant for debug purposes.

## Health checks configuration

| Name                   | Type    | Default             | Comment                                       |
| ---------------------- | ------- | ------------------- | --------------------------------------------- |
| APP_HEALTH_ENABLED     | Boolean | false               | Serve health checks                           |
| APP_HEALTH_PORT        | Number  | 4000                | Port on which serve the health checks         |
| APP_HEALTH_ALLOWED     | String  | ::1/128,127.0.0.0/8 | Allowed CIDR for the health checks            |
| APP_HEALTH_WORKER_BEAT | String  |                     | URL for worker beat pushes                    |
| APP_LOG_WORKER_BEAT    | Boolean | true                | Wether or not to log worker beat console logs |

## Session configuration

| Name                 | Type   | Default | Comment                                        |
| -------------------- | ------ | ------- | ---------------------------------------------- |
| APP_SESSION_SECRET   | String |         | Secret use for JWT signature on session tokens |
| APP_SESSION_LIFETIME | String | 1h      | Session token lifetime                         |

## Storage configuration

| Name                                     | Type                         | Default        | Comment                                                                                                                |
| ---------------------------------------- | ---------------------------- | -------------- | ---------------------------------------------------------------------------------------------------------------------- |
| APP_STORAGE_ENDPOINT                     | String                       |                | API endpoint for the Object storage                                                                                    |
| APP_STORAGE_ACCESS_KEY                   | String                       |                | Access Key for object storage                                                                                          |
| APP_STORAGE_SECRET_KEY                   | String                       |                | Secret Key for object storage                                                                                          |
| APP_STORAGE_SSL                          | Boolean                      | true           | Use TLS with the object storage                                                                                        |
| APP_STORAGE_PORT                         | Number                       |                | Port on which the API endpoint is served                                                                               |
| APP_STORAGE_REGION                       | String                       | ap-southeast-1 | Object storage region                                                                                                  |
| APP_STORAGE_BUCKET                       | String                       | app            | Bucket name for the object storage                                                                                     |
| APP_RETENTION_ENABLED                    | Boolean                      | false          | Enable Retention for Screenshot object                                                                                 |
| APP_IMAGE_RETENTION_MODE                 | "GOVERNANCE" or "COMPLIANCE" | GOVERNANCE     | [S3 Object Locking Modes](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html#object-lock-overview) |
| APP_RETENTION_UNTIL                      | Number                       | 30             | Rention period in days                                                                                                 |
| APP_CREATE_BUCKET_IF_NOT_EXISTS          | Boolean                      | false          | Create a bucket if not exists                                                                                          |
| APP_UPDATE_BUCKET_POLICY_ON_EVERY_UPLOAD | Boolean                      | false          | Update Bucket Policy to the latest configs                                                                             |
| APP_RETENTION_CHECK_INTERVAL             | Number                       | 0.5            | Interval in days to check and delete expired screenshots                                                               |

## Sentry configuration

| Name                          | Type    | Default | Comment                               |
| ----------------------------- | ------- | ------- | ------------------------------------- |
| APP_SENTRY_DSN                | String  |         | Sentry DSN                            |
| APP_SENTRY_RELEASE            | String  |         | Sentry version                        |
| APP_SENTRY_ENVIRONMENT        | String  |         | Sentry environment                    |
| APP_SENTRY_TRACING            | Boolean | true    | Enable Sentry tracing                 |
| APP_SENTRY_TRACES_SAMPLE_RATE | Number  | 1.0     | Sentry traces sample rate for tracing |

## Limiter configuration

| Name            | Type   | Default | Comment                                   |
| --------------- | ------ | ------- | ----------------------------------------- |
| APP_LIMITER_API | Number | 1000    | Rate limiter per second per IP on the API |


src/core/config.ts
import chalk from 'chalk';
import { getString, getBoolean, getInteger, getNumber, getPrefix, getStringList, getStringArray } from './env';
import { Dialect } from 'sequelize';
import { RETENTION_MODES } from 'minio';
import { getClientSideFieldLevelEncryptionSettings } from './encryption';

const version = getString('VERSION', '0.0.0-development');

const config = {
    version,

    // miscellaneous
    verbose: getBoolean(getPrefix('VERBOSE'), true),

    jwt: {
        integrated: {
            secret: getString(getPrefix('INTEGRATED_JWT_SECRET_KEY')),
        },
    },

    name: getString(getPrefix('NAME'), 'VORP Monitoring Server'),
    port: getNumber(getPrefix('PORT'), 3000),
    publicPath: getString(getPrefix('PUBLIC_PATH'), '/public/'),
    adminPublicPath: getString(getPrefix('ADMIN_PUBLIC_PATH'), '/public/admin/'),
    manifestFileName: getString(getPrefix('MANIFEST_FILE_NAME'), 'manifest'),
    allowedDomains: getStringArray(getPrefix('ALLOWED_DOMAINS'), []),
    socketClient: getString(getPrefix('SOCKET_CLIENT_DOMAIN')),

    // secure cookies
    cookiePolicy: getString(getPrefix('COOKIE_POLICY'), 'strict'),
    secureCookie: getBoolean(getPrefix('SECURE_COOKIE'), false),

    // health checks
    healthChecks: {
        enabled: getBoolean(getPrefix('HEALTH_ENABLED'), false),
        port: getNumber(getPrefix('HEALTH_PORT'), 4000),
        allowed: getStringList(getPrefix('HEALTH_ALLOWED'), ['::1/128', '127.0.0.0/8']),
        workerBeat: getString(getPrefix('HEALTH_WORKER_BEAT')),
        logWorkerBeat: getBoolean(getPrefix('LOG_WORKER_BEAT'), true),
    },

    // gzip module
    gzip: getBoolean(getPrefix('GZIP'), true),

    // server runtime
    db: {
        uri: getString(getPrefix('DB_URI'), 'mongodb://root:password@localhost:27017'),
        name: getString(getPrefix('DB_NAME'), 'vorp'),
        pool: getInteger(getPrefix('DB_POOL'), 10),

        username: getString(getPrefix('DB_USERNAME'), 'root'),
        password: getString(getPrefix('DB_PASSWORD'), 'password'),

        // Client Side Field Level Encryption (CSFLE)
        // https://docs.mongodb.com/drivers/security/client-side-field-level-encryption-guide/
        encryption: getClientSideFieldLevelEncryptionSettings(),

        // mongocryptd settings
        // only used if encryption is enabled
        cryptd: {
            uri: getString(getPrefix('DB_CRYPTD_URI')),
            mongocryptdBypassSpawn: getBoolean(getPrefix('DB_CRYPTD_BYPASS_SPAWN'), false),
            mongocryptdSpawnArgs: getStringList(getPrefix('DB_CRYPTD_SPWAN_ARGS'), [], ' '),
        },

        integrated: {
            name: getString(getPrefix('INTEGRATED_DB_NAME'), 'app'),
            username: getString(getPrefix('INTEGRATED_DB_USER_NAME'), 'root'),
            password: getString(getPrefix('INTEGRATED_DB_PASSWORD'), 'password'),
            host: getString(getPrefix('INTEGRATED_DB_HOST'), 'localhost'),
            dialect: getString(getPrefix('INTEGRATED_DB_DIALECT'), 'mysql') as Dialect,
            pool: getInteger(getPrefix('INTEGRATED_DB_POOL'), 10),
            port: getInteger(getPrefix('INTEGRATED_DB_PORT'), 3306),
        },
    },

    storage: {
        provider: {
            endPoint: getString(getPrefix('STORAGE_ENDPOINT')),
            accessKey: getString(getPrefix('STORAGE_ACCESS_KEY')),
            secretKey: getString(getPrefix('STORAGE_SECRET_KEY')),
            useSSL: getBoolean(getPrefix('STORAGE_SSL'), true),
            port: getInteger(getPrefix('STORAGE_PORT')),
            region: getString(getPrefix('STORAGE_REGION'), 'ap-southeast-1'),
        },
        bucket: getString(getPrefix('STORAGE_BUCKET'), 'app'),
        createBucketIfNotExists: getBoolean(getPrefix('CREATE_BUCKET_IF_NOT_EXISTS'), false),
        updateBucketPolicyOnUpload: getBoolean(getPrefix('UPDATE_BUCKET_POLICY_ON_EVERY_UPLOAD'), false),
        retention: {
            enabled: getBoolean(getPrefix('SCREENSHOT_RETENTION_ENABLED'), false),
            mode: getString(getPrefix('SCREENSHOT_RETENTION_MODE'), RETENTION_MODES.GOVERNANCE) as RETENTION_MODES,
            until: getNumber(getPrefix('SCREENSHOT_RETENTION_UNTIL'), 30),
            expiresChecksInterval: getNumber(getPrefix('SCREENSHOT_RETENTION_CHECK_INTERVAL'), 0.5),
        },
        compression: {
            width: getNumber(getPrefix('IMAGE_WIDTH'), 1275),
            height: getNumber(getPrefix('IMAGE_HEIGHT'), 750),
            fit: getString(getPrefix('IMAGE_FIT'), 'contain'),
            quality: getNumber(getPrefix('IMAGE_COMPRESSION_QUALITY'), 50),
        },
    },

    redis: {
        uri: getString(getPrefix('REDIS_URI'), 'redis://127.0.0.1:6379'),
    },

    bull: {
        persistJobs: getBoolean(getPrefix('BULL_PERSIST'), false),
        enableMonitor: getBoolean(getPrefix('BULL_MONITOR'), false),
    },

    session: {
        secret: getString(getPrefix('SESSION_SECRET'), 'localSecret'),
        lifetime: getString(getPrefix('SESSION_LIFETIME'), '1h'),
    },

    limiter: {
        api: getNumber(getPrefix('LIMITER_API'), 1000),
    },

    prometheus: {
        enabled: getBoolean(getPrefix('PROMETHEUS_ENABLED'), false),
        internal: getBoolean(getPrefix('PROMETHEUS_INTERNAL'), false),
        internalPort: getInteger(getPrefix('PROMETHEUS_INTERNAL_PORT'), 7788),
        external: getBoolean(getPrefix('PROMETHEUS_EXTERNAL'), false),
        externalPath: getString(getPrefix('PROMETHEUS_EXTERNAL_PATH'), '/metrics'),
        prefix: getString(getPrefix('PROMETHEUS_PREFIX'), 'app_'),
    },

    sentry: {
        org: getString(getPrefix('SENTRY_ORG'), 'vorp'),
        dsn: getString(getPrefix('SENTRY_DSN')),
        release: getString(getPrefix('SENTRY_RELEASE')),
        environment: getString(getPrefix('SENTRY_ENVIRONMENT')),
        tracing: getBoolean(getPrefix('SENTRY_TRACING'), true),
        tracesSampleRate: getNumber(getPrefix('SENTRY_TRACES_SAMPLE_RATE'), 1.0),
    },

    whatsapp: {
        accessToken: getString(getPrefix('WHATSAPP_ACCESS_TOKEN')),
        phoneNumberID: getString(getPrefix('WHATSAPP_PH_NUMBER_ID')),
        webhook: {
            token: getString(getPrefix('WHATSAPP_WEBHOOK_TOKEN')),
        },
    },
};

export type RuntimeConfig = typeof config;

const exitOnError = (error: string): void => {
    console.error(error);
    process.exit(1);
};

export const runValidityChecks = () => {
    if (!config.session.secret) {
        exitOnError(chalk.red('APP_SESSION_SECRET is missing in environment variables'));
    }
};

export default config;
